library/core/src/mem/mod.rs

```diff
@@ -506,6 +512,8 @@ pub const fn align_of_val<T: ?Sized>(val: &T) -> usize {
 ///     - a [slice], then the length of the slice tail must be an initialized
 ///       integer, and the size of the *entire value*
 ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
+///       For the special case where the dynamic tail length is 0, this function
```

Comment:
Are we missing tests exercising this specifically (in the "success" sense, perhaps miri tests or a run-pass test w/ an `assert_eq!`)?

Reply:
This PR adds the relevant test. I don't think there's a useful "success" test here, what do you have in mind?

Reply:
Like, when we're calling `size_of_val_raw` on a struct with a tail who is a slice of size zero or something.

I just think it's weird that we're adding assumptions guaranteeing *safety* here, but adding tests that guarantee an *error*. I think that's because of the note you had above:

> NOTE: the reason this is safe is that if an overflow were to occur already with size 0, then we would stop compilation as even the "statically known" part of the type would already be too big (or the call may be in dead code and optimized away, but then it doesn't matter).

but I'm just confused why a "negative" test is guaranteed for a new "positive" property, if you see what I mean.

Reply:
Yeah I can see how that's confusing. But a test that just checks that `size_of_val_raw` returns the right result for some small slice seems entirely pointless to me, the functionality of that intrinsic is already covered by existing tests. This includes all the Miri tests that use `size_of_val`/`size_of_val_raw`. The key point is that compilation is guaranteed to fail if an overflow were to happen, and that's what the new test ensures.